Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -1642,912 +1344,94 @@ - (BOOL)canCall
 {
-    // Primary language is nil for the emoji keyboard & we want to stay on it after sending
-    if (![self.inputToolbar.contentView.textView.textInputMode primaryLanguage]) {
+    if (self.userLeftGroup) {
+        DDLogDebug(@"%@ Ignoring request to show conversation settings, since user left group", self.tag);
         return;
     }
 
-    // The JSQ event listeners cause a bounce animation, so we temporarily disable them.
-    [self.keyboardController endListeningForKeyboard];
-    [self dismissKeyBoard];
-    [self popKeyBoard];
-    [self.keyboardController beginListeningForKeyboard];
+    OWSConversationSettingsViewController *settingsVC = [OWSConversationSettingsViewController new];
+    settingsVC.conversationSettingsViewDelegate = self;
+    [settingsVC configureWithThread:self.thread];
+    settingsVC.showVerificationOnAppear = showVerification;
+    [self.navigationController pushViewController:settingsVC animated:YES];
 }
 
-#pragma mark - UICollectionViewDelegate
+- (void)didTapTimerInNavbar:(id)sender
+{
+    DDLogDebug(@"%@ Tapped timer in navbar", self.tag);
+    [self showConversationSettings];
+}
 
-// Override JSQMVC
-- (BOOL)collectionView:(JSQMessagesCollectionView *)collectionView
-    shouldShowMenuForItemAtIndexPath:(NSIndexPath *)indexPath
+- (void)collectionView:(JSQMessagesCollectionView *)collectionView
+                             header:(JSQMessagesLoadEarlierHeaderView *)headerView
+    didTapLoadEarlierMessagesButton:(UIButton *)sender
 {
-    if (indexPath == nil) {
-        DDLogError(@"Aborting shouldShowMenuForItemAtIndexPath because indexPath is nil");
-        // Not sure why this is nil, but occasionally it is, which crashes.
-        return NO;
-    }
+    OWSAssert(!self.isUserScrolling);
 
-    // JSQM does some setup in super method
-    [super collectionView:collectionView shouldShowMenuForItemAtIndexPath:indexPath];
+    BOOL hasEarlierUnseenMessages = self.dynamicInteractions.hasMoreUnseenMessages;
 
+    // We want to restore the current scroll state after we update the range, update
+    // the dynamic interactions and re-layout.  Here we take a "before" snapshot.
+    CGFloat scrollDistanceToBottom = self.safeContentHeight - self.collectionView.contentOffset.y;
 
-    // Don't show menu for in-progress downloads.
-    // We don't want to give the user the wrong idea that deleting would "cancel" the download.
-    id<OWSMessageData> message = [self messageAtIndexPath:indexPath];
-    if (message.isMediaMessage && [message.media isKindOfClass:[AttachmentPointerAdapter class]]) {
-        AttachmentPointerAdapter *attachmentPointerAdapter = (AttachmentPointerAdapter *)message.media;
-        return attachmentPointerAdapter.attachmentPointer.state == TSAttachmentPointerStateFailed;
-    }
-
-    // Super method returns false for media methods. We want menu for *all* items
-    return YES;
-}
-
-- (BOOL)collectionView:(UICollectionView *)collectionView
-      canPerformAction:(SEL)action
```

Comment:
Action routing is now done mostly by the cells, so that we can gather concerns about the per-cell actions (share, save, info, delete, copy, etc.)